DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/explicit_data_movement/src/explicit_data_movement.cpp

```diff
@@ -0,0 +1,282 @@
+class ImplicitKernel;
+class ExplicitKernel;
+
+//
+// This version of the kernel demonstrates implicit data movement
+// through SYCL buffers and accessors.
+//
+template<typename T>
+double SubmitImplicitKernel(queue& q, std::vector<T>& in, std::vector<T>& out,
+                            size_t size) {
+  // start the timer
+  auto start = high_resolution_clock::now();
+  
+  {
+    // set up the input and output buffers
+    buffer in_buf(in);
+    buffer out_buf(out);
+
+    // launch the computation kernel
+    auto kernel_event = q.submit([&](handler& h) {
+      accessor in_accessor(in_buf, h, read_only);
+      accessor out_accessor(out_buf, h, write_only, noinit);
+
+      h.single_task<ImplicitKernel>([=]() [[intel::kernel_args_restrict]] {
+        for (size_t  i = 0; i < size; i ++) {
+          out_accessor[i] = in_accessor[i] * i;
+        }
+      });
+    });
+  }
+  
+  // We use the scope above to synchronize the FPGA kernels.
+  // Exiting the scope will cause the buffer destructors to be called
+  // which will wait until the kernel finishes and copy the data back to the
+  // host (if the buffer was written to).
+  // Therefore, at this point in the code, we know the kernels have finished
+  // and the data has been transferred back to the host.
+
+  // stop the timer
+  auto end = high_resolution_clock::now();
+  duration<double, std::milli> diff = end - start;
+
+  return diff.count();
+}
+
+//
+// This version of the kernel demonstrates explicit data movement
+// through explicit USM.
+//
+template<typename T>
+double SubmitExplicitKernel(queue& q, std::vector<T>& in,
+                            std::vector<T>& out, size_t size) {
+  // allocate the device memory
+  T* in_ptr = malloc_device<T>(size, q.get_device(), q.get_context());
+  T* out_ptr = malloc_device<T>(size, q.get_device(), q.get_context());
+
+  // ensure we successfully allocated the device memory
+  if(in_ptr == nullptr) {
+    std::cerr << "ERROR: failed to allocate space for 'in_ptr'\n";
+    return 0;
+  }
+  if(out_ptr == nullptr) {
+    std::cerr << "ERROR: failed to allocate space for 'out_ptr'\n";
+    return 0;
+  }
+
+  // start the timer
+  auto start = high_resolution_clock::now();
+
+  // copy host input data to the device's memory
+  auto copy_host_to_device_event = q.submit([&](handler& h) {
+    h.memcpy(in_ptr, in.data(), size*sizeof(T));
+  });
+
+  // launch the the computation kernel
+  auto kernel_event = q.submit([&](handler& h) {
+    // this kernel must wait until the data is copied from the host's to
+    // the device's memory
+    h.depends_on(copy_host_to_device_event);
+
+    h.single_task<ExplicitKernel>([=]() [[intel::kernel_args_restrict]] {
+      // create device pointers to explicitly inform the compiler these
+      // pointer reside in the device's address space
+      device_ptr<T> in_ptr_d = device_ptr<T>(in_ptr);
+      device_ptr<T> out_ptr_d = device_ptr<T>(out_ptr);
```

Comment:
DRY (Don't Repeat Yourself)
```suggestion
      auto in_ptr_d = device_ptr<T>(in_ptr);
      auto out_ptr_d = device_ptr<T>(out_ptr);
```
or
```suggestion
      device_ptr<T> in_ptr_d { in_ptr };
      device_ptr<T> out_ptr_d { out_ptr };
```